epigraphhub/models/ngboost_models.py

```diff
@@ -0,0 +1,740 @@
+    params smooth: It determines if data will be smoothed or not
+    params ini_date: Determines the beggining of the train dataset
+
+    returns: Dataframe with the predictions for all the cantons
+    """
+
+    df_all = pd.DataFrame()
+
+    # compute the clusters
+    clusters, all_regions, fig = compute_clusters(
+        "switzerland",
+        "cases",
+        ["datum", '"geoRegion"', "entries"],
+        t=0.3,
+        drop_georegions=["CH", "FL", "CHFL"],
+        plot=False,
+    )
+
+    for cluster in clusters:
+        # getting the data
+        df = get_cluster_data(predictors, list(cluster), vaccine=vaccine, smooth=smooth)
+        # filling the nan values with 0
+        df = df.fillna(0)
+
+        for canton in cluster:
+            # apply the model
+
+            target_name = f"{target_curve_name}_{canton}"
+
+            horizon = 14
+            maxlag = 14
+
+            # get predictions and forecast
+
+            df_pred = rolling_predictions(
+                target_name,
+                df,
+                ini_date=ini_date,
+                split=0.75,
+                horizon_forecast=horizon,
+                maxlag=maxlag,
+            )
+
+            df_all = pd.concat([df_all, df_pred])
+
+    return df_all
+
+
+def training_model(target_name, data, ini_date, horizon_forecast=14, maxlag=14):
+
+    # print(data.index[-1])
+    data = data.iloc[:-3]
+
+    # print(data.index[-1])
+    target = data[target_name]
+
+    df_lag = build_lagged_features(copy.deepcopy(data), maxlag=maxlag)
+
+    ini_date = max(
+        df_lag.index[0], target.index[0], datetime.strptime(ini_date, "%Y-%m-%d")
+    )
+
+    df_lag = df_lag[ini_date:]
+    target = target[ini_date:]
+    target = target.dropna()
+    df_lag = df_lag.dropna()
+
+    # condition to only apply the model for datas with values different of 0
+    if np.sum(target) > 0.0:
+
+        # remove the target column and columns related with the day that we want to predict
+        df_lag = df_lag.drop(data.columns, axis=1)
+
+        # targets
+        targets = {}
+
+        for T in np.arange(1, horizon_forecast + 1, 1):
+            if T == 1:
+                targets[T] = target.shift(-(T - 1))
+            else:
+                targets[T] = target.shift(-(T - 1))[: -(T - 1)]
+                # print(T, len(df_lag), len(fit_target))
+                # print(df_lag.index,fit_target.index)
+
+        models = []
+
+        X_train = df_lag.iloc[:-1]
+
+        for T in range(1, horizon_forecast + 1):
+            # training of the model with all the data available
+
+            # print(T)
+
+            tgt = targets[T][: len(X_train)]
+
+            i = 0
+
+            while i < len(tgt):
+                if tgt[i] <= 0:
+
+                    tgt[i] = 0.01
+                i = i + 1
+
+            # if not os.path.exists(f'../opt/models/saved_models/ml/ngboost_{target_name}_{T}D.joblib'):
```

Comment:
if you need to keep the comment, please add a note about how it is commented.
another point: void paths like this inside the function, probably it is better to get the path from a variable or from an environment variable

Reply:
@eduardocorrearaujo and @xmnlab you normally don´t edit `poetry.lock` manually, in this case just run `poetry add joblib` and it will update the `pyproject.toml` and `poetry.lock` as well.

Reply:
yep, I just learned about `poetry add` yesterday. thanks for sharing!